### PR TITLE
Add support for compute capability in IR Lowering

### PR DESCRIPTION
**Pull request by testhound**
This PR addresses issue 7954: https://github.com/numba/numba/issues/7954 by adding compute capability to the compiler flags which can be checked during lowering. It also adds a test case where tanh is converted to tanh.approx.f32 when fastmath is enabled and compute capability >= 7.5.

One consequence of this change is that several test case had to be converted to CUDATestCase because of the driver call in compile_cuda to get the compute capability.

### CODE REVIEW

**Comment by testhound** [comment]
gpuci run tests

**Review by gmarkall — CHANGES_REQUESTED** [review]
The fact that the tests of `compile_ptx` needed to be converted to `CUDATestCase` subclasses  suggests an issue - compilation to PTX shouldn't require a CUDA device to be present / cause any interaction with the driver API.

Before this can proceed, it needs to be changed such that `compile_ptx` doesn't cause any interactions with the driver / hardware.

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
@gmarkall fixed so that `compile_ptx` does not interact with the driver. I did have to convert `TestCudaMandel `to derive from `CUDATestCase` because set device=true in the jit decorator.

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Review by gmarkall — CHANGES_REQUESTED** [review]
Many thanks for the PR! I have made a small suggestion on the diff about the flag type (I also think the `nvvm_options_type` which probably inspired it should go, but that's something for another PR). I have a couple of other concerns:

The addition of use of the default compute capability in `compile_cuda()` now means the default is potentially used in two places (there and in `compile_ptx()`), which is a bit harder to reason about the safety of. I think it'd be safer if the compute capability always had to be passed in to `compile_cuda()`, then the default only need be used for `compile_ptx()`. I think the following change would achieve this:

```diff
diff --git a/numba/cuda/compiler.py b/numba/cuda/compiler.py
index 664d82c8c..99c5b7e2d 100644
--- a/numba/cuda/compiler.py
+++ b/numba/cuda/compiler.py
@@ -191,6 +191,9 @@ class CUDACompiler(CompilerBase):
 def compile_cuda(pyfunc, return_type, args, debug=False, lineinfo=False,
                  inline=False, fastmath=False, nvvm_options=None,
                  cc=None):
+    if cc is None:
+        raise ValueError('Compute Capability must be supplied')
+
     from .descriptor import cuda_target
     typingctx = cuda_target.typing_context
     targetctx = cuda_target.target_context
@@ -215,7 +218,7 @@ def compile_cuda(pyfunc, return_type, args, debug=False, lineinfo=False,
         flags.fastmath = True
     if nvvm_options:
         flags.nvvm_options = nvvm_options
-    flags.compute_capability = cc or config.CUDA_DEFAULT_PTX_CC
+    flags.compute_capability = cc
 
     # Run compilation pipeline
     from numba.core.target_extension import target_override
@@ -277,6 +280,7 @@ def compile_ptx(pyfunc, args, debug=False, lineinfo=False, device=False,
         'opt': 3 if opt else 0
     }
 
+    cc = cc or config.CUDA_DEFAULT_PTX_CC
     cres = compile_cuda(pyfunc, None, args, debug=debug, lineinfo=lineinfo,
                         fastmath=fastmath,
                         nvvm_options=nvvm_options, cc=cc)
@@ -292,7 +296,6 @@ def compile_ptx(pyfunc, args, debug=False, lineinfo=False, device=False,
         lib, kernel = tgt.prepare_cuda_kernel(cres.library, cres.fndesc, debug,
                                               nvvm_options, filename, linenum)
 
-    cc = cc or config.CUDA_DEFAULT_PTX_CC
     ptx = lib.get_asm_str(cc=cc)
     return ptx, resty
 
```
What do you think of this suggestion?

My second question is - what happens with the `tanh` implementation when used in a function compiled with `compile_ptx`? Can you add a test that compiles for two different CCs (one less than and one at least 7.5) please?

**Comment by testhound** [comment]
@gmarkall I incorporated all your requested changes.

**Comment by testhound** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
> My second question is - what happens with the tanh implementation when used in a function compiled with compile_ptx? Can you add a test that compiles for two different CCs (one less than and one at least 7.5) please?

Can you add the tests of `compile_ptx` please?

**Comment by testhound** [comment]
> > My second question is - what happens with the tanh implementation when used in a function compiled with compile_ptx? Can you add a test that compiles for two different CCs (one less than and one at least 7.5) please?
> 
> Can you add the tests of `compile_ptx` please?

@gmarkall I added a new test "test_tanhf_no_fastmath" that is skipped if compute capability is < 7.5.

File: numba/cuda/mathimpl.py [hunk 1]
```diff
@@ -249,13 +248,53 @@ def lower_ldexp_impl(context, builder, sig, args):
...
 
+def impl_tanh(ty, libfunc):
+    def lower_tanh_impl(context, builder, sig, args):
+        def get_compute_capability():
+            cstk = targetconfig.ConfigStack()
+            if cstk:
+                flags = cstk.top()
+                return flags.compute_capability
+            else:
+                return None
```
Comment (gmarkall) [inline numba/cuda/mathimpl.py:259]: From OOB discussion, I think the conclusion is that the top of the config stack can never be `None` (an `IndexError` would be raised in this case).

```suggestion
            flags = targetconfig.ConfigStack().top()
            return flags.compute_capability
```

**Review by gmarkall — CHANGES_REQUESTED** [review]
Following some further OOB discussion, it seems like the remaining step is to add a test of compiling `tanh` with and without fastmath using `compile_ptx()`, specifying compute capabilities 7.0 in one instance and 7.5 in another - this will mean that both paths through the tanh implementation get tested on all systems, regardless of the device present in the machine.

This should also ensure that there isn't a case we've missed where the config stack is empty when `compile_ptx()` is called - I don't think there is one, bit I think the config stack stuff is very hard to reason about due to its use of TLS and hidden / global state.

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
@gmarkall new testcase added as you requested.

**Comment by testhound** [comment]
@gmarkall fixed merge conflicts and waiting on your review.

**Review by gmarkall — CHANGES_REQUESTED** [review]
Many thanks for the updates - some comments on the diff, but on the whole I think this is almost done.

**Comment by testhound** [comment]
@gmarkall all request changes made to the PR.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
@testhound what about enabling `test_tanhf_compile_ptx` to run regardless of the CC of the installed device?

**Comment by testhound** [comment]
@gmarkall okay I was planning to try that next, I misunderstood you comment and applied it to test_tanhf.

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
@gmarkall the gpuci test failed in a way unrelated to my latest change.

**Review by gmarkall — APPROVED** [review]
I think this looks good now - many thanks for the updates!

**Pull request merged**